# A process moves on the integers 1, 2, 3, 4, and 5

# It starts at 1 and, on each successive step, moves to an integer greater than its present position, moving with equal probability to each of the remaining larger integers

# State 5 is an absorbing state

# Find the expected number of steps to reach state 5

_____

# The transition matrix for this process is:

# $P = \begin{pmatrix}0 & 1/4 & 1/4 & 1/4 & 1/4\\ 0 & 0 & 1/3 & 1/3 & 1/3\\ 0 & 0 & 0 & 1/2 & 1/2\\ 0 & 0 & 0 & 0 & 1\\ 0 & 0 & 0 & 0 & 1\end{pmatrix}$

### Lucky for us, this is already in its canonical form

# $\implies Q = \begin{pmatrix}0 & 1/4 & 1/4 & 1/4\\ 0 & 0 & 1/3 & 1/3\\ 0 & 0 & 0 & 1/2\\ 0 & 0 & 0 & 0\end{pmatrix}$

# $\implies (I-Q) = \begin{pmatrix}1 & -1/4 & -1/4 & -1/4\\ 0 & 1 & -1/3 & -1/3\\ 0 & 0 & 1 & -1/2\\ 0 & 0 & 0 & 1\end{pmatrix}$

### Using numpy to calculate the inverse

In [1]:
import numpy as np

In [2]:
matrix = np.array([[1, -0.25, -0.25, -0.25],
                  [0, 1, -1/3.0, -1/3.0],
                  [0, 0, 1, -0.5],
                  [0, 0, 0, 1]])

In [4]:
N = np.linalg.inv(matrix)
N

array([[ 1.        ,  0.25      ,  0.33333333,  0.5       ],
       [ 0.        ,  1.        ,  0.33333333,  0.5       ],
       [ 0.        ,  0.        ,  1.        ,  0.5       ],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])

# $\implies N = \begin{pmatrix}1 & 1/4 & 1/3 & 1/2\\ 0 & 1 & 1/3 & 1/2\\ 0 & 0 & 1 & 1/2\\ 0 & 0 & 0 & 1\end{pmatrix}$

# $c = \begin{pmatrix}1\\ 1\\ 1\\ 1\end{pmatrix}$

### Using numpy to do this matrix math

In [10]:
c = np.ones(4)[:, np.newaxis]

In [13]:
Nc = np.matmul(N, c)
Nc

array([[ 2.08333333],
       [ 1.83333333],
       [ 1.5       ],
       [ 1.        ]])

# $\implies Nc = \begin{pmatrix}2.08333\\ 1.83333\\ 1.5\\ 1\end{pmatrix}$

# Recall: $Nc$ tells us the time to absorption if we start in each state

### Since we can only start in start 1, we only care about the top value

# Therefore, the expected number of steps to reach stage 5 is 2.083

_____

## We can confirm this by simulation

In [29]:
def sim():
    list_states = [1,2,3,4,5]
    
    state = 1
    
    n_steps = 0
    
    while state < 5:
        n_steps += 1
        index = list_states.index(state)
        list_states = list_states[index+1:]
        state = np.random.choice(list_states)
    return n_steps

In [32]:
results = []
for i in range(100000):
    results.append(sim())

In [33]:
np.mean(results)

2.0833699999999999

### Bang on